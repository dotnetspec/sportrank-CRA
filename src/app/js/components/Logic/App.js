import Header from '../UI/Header'
import Main from './Main'
import React, {
  useState,
  useEffect
} from 'react';
import {
  withRouter
} from 'react-router-dom';
import JSONops from './JSONops';
import {
  _loadsetJSONData,
  getNewRankId,
  getDefaultRankingList
} from '../SideEffects/io/Jsonio';
import {
  _loadExternalBalance
  //,
  // _mapCurrentUserAccounts,
  // mapTheAccounts
} from '../SideEffects/io/web3io';
import web3 from '../../../../web3';
import DSportRank from '../../../../ABIaddress';
import ChangeState from './ChangeState';

/**
 * Functional component representing the highest order component. Any user
 * updates in child components should trigger an event in this
 * class so that the current user details can be re-fetched from
 * the contract and propagated to all children that rely on it
 *
 * @extends React.Component
 */
export function App({
  props
}) {
  const [user, setUser] = useState({});
  const [userName, setUserName] = useState('');
  const [account, setAccount] = useState('');
  const [error, setError] = useState({});
  const [userAccounts, setuserAccounts] = useState([])
  const [balance, setBalance] = useState(0)
  const [data, setdata] = useState([])
  const [updatedExtAcctBalCB, setupdatedExtAcctBalCB] = useState(0)
  const [isLoading, setIsLoading] = useState(true);
  const [userRankingLists] = useState([]);
  const [isUserInJson, setIsUserInJson] = useState(false);
  const [isLoadingJSON, setisLoadingJSON] = useState(true);
  const [isCurrentUserActive, setIsCurrentUserActive] = useState(false);
  const [isCurrentUserActiveCB] = useState(false);
  const [isRankingIDInvalid] = useState(false);
  const [newrankId, setnewrankId] = useState('');
  const [rankingListData, setrankingListData] = useState([]);
  const [viewingOnlyCB, setviewingOnlyCB] = useState(true);
  const [contactno, setcontactno] = useState('');
  const [email, setemail] = useState('');
  const [description, setdescription] = useState({});
  const [specificRankingOptionBtns, setspecificRankingOptionBtns] = useState(false);
  const [rank, setrank] = useState('1');
  const [ranknameHasChanged, setranknameHasChanged] = useState(false);
  const [resultInfoForDisplay, setResultInfoForDisplay] = useState('');
  //REVIEW: set as a global var. Perhaps change to environment var ?:
  const rankingDefault = '5c36f5422c87fa27306acb52';

  function _loadExternalBalance_callback(externalbalance) {
    if (externalbalance !== undefined) {
      setupdatedExtAcctBalCB(externalbalance);
    }
  }

  //Below appears to be relevant to user events not e.g. callbacks that fetch data
  const setspecificRankingOptionBtnsCB = () => {
    setspecificRankingOptionBtns(true);
  }
  const handleListAllChildClick = () => {
    setspecificRankingOptionBtns(false);
  }
  //cb from GlobalRankings.js to set the rank state as view only
  //REVEIW: is this necessary? wasn't working until noticed it ...
  const setOnCallbackviewingOnlyCB = (viewingOnlyCBval) => {
    setviewingOnlyCB(viewingOnlyCBval);
  }
  //cb from PlayerStatusBtn.js to set the state of the button
  const setOnCallbackisCurrentUserActiveCB = (btnState) => {
    setIsCurrentUserActive(btnState);
  }
  //in GlobalRankingViewBtn
  const setnewrankIdCB = (newrankIdtxt) => {
    setnewrankId(newrankIdtxt);
    setranknameHasChanged(true);
    setIsLoading(false);
    _loadsetJSONData(newrankIdtxt, setrankingJSONdataCB);
  }
  const setResultInfoForDisplayCB = (text) => {
    console.log('setResultInfoForDisplayCB in app.js', text);
    setResultInfoForDisplay(text);
  }
  //cb from createuser.js to set the username
  //in time for getNewRankingID() to put it in the json
  const setuserNameCB = (name) => {
    setUserName(name);
    //this was originally
    //may need to account for this
    //setUser(name);
  }
  const setcontactNoCB = (number) => {
    setcontactno(number);
  }
  const setemailCB = (oppoEmailTxt) => {
    setemail(oppoEmailTxt);
  }
  const setuserDescCB = (userObj, description) => {
    userObj.description = description;
    setUser(userObj);
  }

  const setuserCB = (userObj, name, contactno, email, description) => {
    userObj.username = name;
    userObj.contactno = contactno;
    userObj.email = email;
    userObj.description = description;
    setUser(userObj);
  }

  const setrankingJSONdataCB = (datatoSet) => {
    setdata(datatoSet);
    setisLoadingJSON(false);
    setIsUserInJson(JSONops.isPlayerListedInJSON(data, user.username));
    setrank(JSONops._getUserValue(data, user.username, "RANK"));
    setIsCurrentUserActive(JSONops._getUserValue(data, user.username, "ACTIVE"));
  }
  //#endregion
  //#region Helper methods
  //REVIEW: Possible to getUserRank in App.js (and set state) rather than Home.js?
  //currently no - problem is waiting for username to check against ran
  /**
   * Sets the App state error and redirects the user to the error page
   *
   * @param {Error} err - error encountered
   */
  const _onError = (err, source) => {
    if (source) err.source = source;
    //setState({ error: err });
    setError(err);
    props.history.push('/whoopsie');
  }
  //#endregion

  const processStateAfter_loadCurrentUserAccounts = (userAcctArr) => {
    console.log('userAcctArr before clean', userAcctArr)
    userAcctArr = ChangeState.cleanUpUserSRAccountData(userAcctArr);
    console.log('userAcctArr after clean', userAcctArr)
    if (userAcctArr) {
      setuserAccounts(userAcctArr);
      //console.log('userAcctArr[0].userAccount', userAcctArr[0])
      //REVIEW: sort out this duplication
      setAccount(userAcctArr[0]);
      setUser(userAcctArr[0])
      setError(userAcctArr[0].error);
      setUserName(userAcctArr[0].username);
      console.log(userAcctArr[0])
      if (userAcctArr[0].data !== undefined) {
        setIsUserInJson(JSONops.isPlayerListedInJSON(userAcctArr[0].data, userAcctArr[0].username));
        setIsCurrentUserActive(JSONops._getUserValue(userAcctArr[0].data, userAcctArr[0].username, "ACTIVE"));
      } else {
        setIsUserInJson(false);
      }
      setnewrankId(userAcctArr[0].newrankId);
      setcontactno(userAcctArr[0].contactno);
      setemail(userAcctArr[0].email);
      setdescription(userAcctArr[0].description);
      setBalance(userAcctArr[0].balance);
      setviewingOnlyCB(true);
    } else {
      console.log('user undefined')
    }
  }
  //#region React lifecycle events
  useEffect(() => {
    setIsLoading(true);
    async function fetchData() {
      //from the Blockchain via web3io:
      //await _loadExternalBalance(_loadExternalBalance_callback);

      await getDefaultRankingList(rankingDefault, getDefaultRankingList_callback);
      async function getDefaultRankingList_callback(json) {
        setrankingListData(json);
      }

      const mapTheAccounts = async (accountsArray) => {
        web3.eth.getAccounts()
        .then(function(expectingAResult){
            return expectingAResult;
        })
        .then(async function(accountsArray){
          //based on: https://flaviocopes.com/javascript-async-await-array-map/
          //get the hashes by async
          //each separate piece of async data needs these parts ...
          const functionWithPromise = item => { //a function that returns a promise
            return Promise.resolve(DSportRank.methods.owners(item).call())
          }

          const anAsyncFunction = async item => {
            return await functionWithPromise(item)
          }
          //wait for the async hash calls to resolve for all accounts
          const getData = async () => {
            return await Promise.all(accountsArray.map(item => anAsyncFunction(item)))
          }
          const data = await getData()
          //each separate piece of async data needs these parts (above)  ...

          //use the hashes to get the user data from contract by async
          const functionGetUserDataWithPromise = item => { //a function that returns a promise
            return Promise.resolve(DSportRank.methods.users(item).call());
          }

          const anAsyncFunctionToGetUserData = async item => {
            return await functionGetUserDataWithPromise(item)
          }
          //wait for the async hash calls to resolve for all users
          const getUserData = async () => {
            return await Promise.all(data.map(item => anAsyncFunctionToGetUserData(item)))
          }

          const userdata = await getUserData()
          console.log('user data in the list', userdata)

          const functionWithPromiseToGetBal = item => { //a function that returns a promise
            return Promise.resolve(web3.eth.getBalance(item.owner));
          }

          const anAsyncFunctionToGetBal = async item => {
            return await functionWithPromiseToGetBal(item)
          }
          //wait for the async hash calls to resolve for all accounts
          const getDataWithBal = async () => {
            return await Promise.all(userdata.map(item => anAsyncFunctionToGetBal(item)))
          }
          const userdataWithBal = await getDataWithBal()

          //var numbers = [4, 9, 16, 25];
          //var userdataWithBalArr = userdata.map(item => web3.eth.getBalance('0xAC5491BB066c98fec13046928a78761c0B1E5603'));
          console.log('userdataWithBal', userdataWithBal)
          return userdata;
        }).then(function(resolvedUserData){
                  processStateAfter_loadCurrentUserAccounts(resolvedUserData);
              }).catch(function(error) {
                   console.log('error is:', error)
              }).then(function() {
                   console.log('in the last then')
              });
            }
        await mapTheAccounts();

        setIsLoading(false);
    }
    fetchData();
  }, []); // Or [someId] (sent as a param to a function) if effect needs props or state (apparently)

  if (!isLoading) {
    console.log('userName', userName)
    return ( <
      div >
      <
      Header data-testid = 'header'
      user = {
        user
      }
      username = {
        userName
      }
      account = {
        account
      }
      description = {
        description
      }
      setuserDescCB = {
        setuserDescCB
      }
      userAccounts = {
        userAccounts
      }
      balance = {
        balance
      }
      error = {
        error
      }
      isCurrentUserActive = {
        isCurrentUserActive
      }
      isCurrentUserActiveCB = {
        isCurrentUserActiveCB
      }
      setOnCallbackisCurrentUserActiveCB = {
        setOnCallbackisCurrentUserActiveCB
      }
      setspecificRankingOptionBtnsCB = {
        setspecificRankingOptionBtnsCB
      }
      onListAllChildClick = {
        handleListAllChildClick
      }
      specificRankingOptionBtns = {
        specificRankingOptionBtns
      }
      // onAfterUserUpdate = {
      //   mapTheAccounts
      // }
      onError = {
        _onError
      }
      rankingJSONdata = {
        data
      }
      rankingListJSONdata = {
        rankingListData
      }
      updatedExtAcctBalCB = {
        updatedExtAcctBalCB
      }
      usersRankingLists = {
        userRankingLists
      }
      isUserInJson = {
        isUserInJson
      }
      rankingDefault = {
        rankingDefault
      }
      newrankId = {
        newrankId
      }
      setnewrankId = {
        setnewrankId
      }
      resultInfoForDisplay = {
        resultInfoForDisplay
      }
      setResultInfoForDisplayCB = {
        setResultInfoForDisplayCB
      }
      />

      <
      Main data-testid = 'main'
      user = {
        user
      }
      setuserCB = {
        setuserCB
      }
      username = {
        userName
      }
      setuserNameCB = {
        setuserNameCB
      }
      contactno = {
        contactno
      }
      setcontactNoCB = {
        setcontactNoCB
      }
      email = {
        email
      }
      setemailCB = {
        setemailCB
      }
      description = {
        description
      }
      setuserDescCB = {
        setuserDescCB
      }
      resultInfoForDisplay = {
        resultInfoForDisplay
      }
      setResultInfoForDisplayCB = {
        setResultInfoForDisplayCB
      }
      account = {
        account
      }
      userAccounts = {
        userAccounts
      }
      error = {
        error
      }
      setspecificRankingOptionBtnsCB = {
        setspecificRankingOptionBtnsCB
      }
      specificRankingOptionBtns = {
        specificRankingOptionBtns
      }

      // onAfterUserUpdate = {
      //   _mapCurrentUserAccounts
      // }
      onError = {
        _onError
      }
      rankingJSONdata = {
        data
      }
      setrankingJSONdataCB = {
        setrankingJSONdataCB
      }
      rankingListJSONdata = {
        rankingListData
      }
      rank = {
        rank
      }
      ranknameHasChanged = {
        ranknameHasChanged
      }
      isCurrentUserActive = {
        isCurrentUserActive
      }
      setOnCallbackisCurrentUserActiveCB = {
        setOnCallbackisCurrentUserActiveCB
      }
      isRankingIDInvalid = {
        isRankingIDInvalid
      }
      newrankId = {
        newrankId
      }
      setnewrankIdCB = {
        setnewrankIdCB
      }
      getNewRankingID = {
        getNewRankId
      }
      rankingDefault = {
        rankingDefault
      }
      viewingOnlyCB = {
        viewingOnlyCB
      }
      setviewingOnlyCB = {
        setOnCallbackviewingOnlyCB
      }
      isUserInJson = {
        isUserInJson
      }
      loadingJSON = {
        isLoadingJSON
      }
      />

      <
      /div>
    );
  } else {
    return ( <
      div data-testid = 'loading' > Loading... < /div>
    );
  }
  //}
  //#endregion
}

export default withRouter(App)
