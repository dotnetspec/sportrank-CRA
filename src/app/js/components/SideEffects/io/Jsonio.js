import axios from 'axios';
//TODO: all functions using _sendJSONData will need to be updated to use this
//one that includes the rankingID
  //_sendJSONDataWithRankingID: function(data, rankingID){
  // export function Jsonio({
  //   props
  // }) {
    export async function _sendJSONDataWithRankingID (data, rankingID) {
    console.log('rankingID inside _sendJSONDataWithRankingID',rankingID)
    console.log('data inside _sendJSONDataWithRankingID',data)
    //console.log('inside _sendJSONDataWithRankingID')
    let httpString = "https://api.jsonbin.io/b/";
    //httpString += rankingID + '"';
    httpString += rankingID;
    let req = new XMLHttpRequest();

        req.onreadystatechange = () => {
          if (req.readyState === XMLHttpRequest.DONE) {
            console.log('httpString in req.onreadystatechange', httpString);
            //NB. when checking on jsonbin.io e.g. https://jsonbin.io/5c340b667b31f426f8531274/1
            //ensure you include the version number to see that the array has been 'PUT'
            console.log('req.responseText in _sendJSONDataWithRankingID', req.responseText);
            //console.log(req.responseText);
            return req.responseText;
          }
        };
        //NOTE: it is the api.jsonbin NOT the jsonbin.io!
        //JSON data can and should be in ANY order
        //bin id is: https://jsonbin.io/5bd82af2baccb064c0bdc92a/
        //use above to edit manually.
        //to view latest https://api.jsonbin.io/b/5bd82af2baccb064c0bdc92a/latest

        req.open("PUT", httpString, true);
        req.setRequestHeader("Content-type", "application/json");
        let myJsonString = JSON.stringify(data);
        console.log('httpString, data in _sendJSONDataWithRankingID', httpString, data);

        //console.log('data.id alone in _sendJSONDataWithRankingID', data.id)
        //I think data.id will only be defined IF this is a new ranking
        //if this is a new ranking send an array, not just an object
        //if this is a new ranking id will be 1
        //HACK: there may be a better way to test that this is a new ranking and user
        //the first entry to jsonbin must have array brackets so that responseJson can be
        //correctly displayed in BootstrapTable
        if(data.id === 1){
        const myJsonStringAsArray = "[" + myJsonString + "]";
          req.send(myJsonStringAsArray);
        }else{
          req.send(myJsonString);
        }
        //return null;
}


//Reference:
//https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch
//REVIEW: Possibly implement requestConfig later ...

const requestConfig = {
        method: 'GET',
        mode: 'cors',
        headers: {
          'Content-Type': 'application/json'
        }
        // ,
        // body: JSON.stringify({
        //   'test',
        //   'test'
        // })
      };


 export async function getDefaultRankingList (rankingDefault, getDefaultRankingList_callback) {
      //export const getDefaultRankingList = async (rankingDefault, getDefaultRankingList_callback) => {
        try {
          let httpStr = 'https://api.jsonbin.io/b/' + rankingDefault + '/latest';
          axios.get(httpStr)
          .then(res => {
            const json = res.data;
            //console.log('json', json)
            getDefaultRankingList_callback(json)
            //return checkUndefined(json);
            //return json;
            //setrankingListData(json);
          })
        } catch (e) {
          //setState({ error: e });
          //setError(e);
          return e;
        } finally {

        }
      }

//asyncFetch abstracted to enable mocking
 export async function asyncFetch(url) {
  return await fetch(url);
}

 export async function _loadsetJSONData (newrankId, _loadsetJSONData_callback){
  console.log('newrankId IN _loadsetJSONData', newrankId);
      try {
            let httpStr = 'https://api.jsonbin.io/b/' + newrankId + '/latest';

            await fetch(httpStr)
               .then((response) => response.json())
               .then((responseJson) => {
                 if(responseJson.length !== 0){
                   responseJson = checkUndefined(responseJson);
                   _loadsetJSONData_callback(responseJson);
                     }
               })
            }catch (err) {
               return console.error(err);
            }
}

 export async function _loadsetRankingListJSONData (rankingDefault, _loadsetRankingListJSONData_callback){
//REVIEW: following is attempt to refactor fetchjson
//await fetchJSON(rankingDefault, _loadsetRankingListJSONData_callback);
  try {
          let httpStr = 'https://api.jsonbin.io/b/' + rankingDefault + '/latest';
          await fetch(httpStr, requestConfig)
           .then((response) => response.json())
           .then((responseJson) => {
             if(responseJson.length !== 0){
               responseJson = checkUndefined(responseJson);
               _loadsetRankingListJSONData_callback(responseJson);
              }
           })
        }catch (err) {
             return console.error(err);
          }
}

//get a new rankid ready in case user wants/needs to create a new ranking
//do this after _loadsetJSONData so that we will already have the correct username
//getNewRankId = async () => {
 export async function getNewRankId(description, account, email, contactno, user, getNewRankId_callback) {
  //console.log('userNameCB in getNewRankId in app', this.state.userNameCB)
    try{
    //this.setState({ isLoading: true});
    let req = new XMLHttpRequest();
      req.onreadystatechange = () => {
        //this async section tests whether the result
        //from the code lower down has been returned
        //(without using await)
        if (req.readyState === XMLHttpRequest.DONE) {

          const resulttxt = JSON.parse(req.responseText);
          getNewRankId_callback(resulttxt);
          //only here can set state (once result is back)
          //this.setState({ newrankId: resulttxt.id});
          //this.setState({ ranknameHasChanged: true});
          //this.setState({ isLoading: false});
        }
      };
      //NB: following will send the request but
      //need to wait for the results to come back
      //(above) before any further processing can be
      //done
      var obj = {
      DATESTAMP: Date.now(),
      ACTIVE: true,
      DESCRIPTION: description,
      CURRENTCHALLENGERNAME: "AVAILABLE",
      CURRENTCHALLENGERID: 0,
      ACCOUNT: account,
      EMAIL: email,
      CONTACTNO: contactno,
      RANK: 1,
      NAME: user,
      id: 1 };

      let myJSON = JSON.stringify(obj);
      //console.log('getNewRankId using myJSON', myJSON)

      req.open("POST", "https://api.jsonbin.io/b", true);
      //req.open("PUT", "https://api.jsonbin.io/b", true);
      req.setRequestHeader("Content-type", "application/json");
      //req.send('{"Player": "Johan Straus"}') || {}
      req.send(myJSON)
      //|| {}
      }catch (err) {
      // stop loading state and show the error
      //console.log(err)
      //this.setState({ isLoading: false, error: err.message });
    };
      return null;
  }

//NB: this function cannot actually be used in app.js as it's only simulating the one
//nested inside the mapping function of _loadCurrentUserAccounts
//    async function _loadCurrentUserAccountsInsideMapping(address, _loadCurrentUserAccountsInsideMapping_callback){
//     // var MyContract = web3.eth.contract(ABI);
//     // var contractInstance = MyContract.at(address);
//     //const contractInstance = new DSportRank();
//     // window.addEventListener('load', function() {
//     //   const web3 = new Web3(Web3.givenProvider || 'ws://localhost:8546', null, {});
//     // })
//     //console.log('web3.currentProvider', web3.currentProvider);
//     const usernameHash = await DSportRank.methods.owners(address).call();
//
//     // const myContract = new web3.eth.Contract([...], '0xde0B295669a9FD93d5F28D9Ec85E40f4cb697BAe', {
//     // defaultAccount: '0x1234567890123456789012345678901234567891', // default from address
//     // defaultGasPrice: '20000000000' // default gas price in wei, 20 gwei in this case
// //});
//     // if(!DSportRank.isConnected()) {
//     //
//     //     console.log('not connected to web3!')
//     //
//     // } else {
//       //const usernameHash = await DSportRank.methods.owners(address).call();
//           // console.log('got usernameHash after await', usernameHash)
//           // console.log('exec at', executingAt());
//           // get user details from contract
//       const user = await DSportRank.methods.users(usernameHash).call();
//       //const user = "player1";
//       const balance = 2;
//       // gets the balance of the address
//       // let balance = await web3.eth.getBalance(address);
//       // balance = web3.utils.fromWei(balance, 'ether');
//       // console.log('balance', balance);
//       const contractObj = {address: address, user: user, balance: balance};
//       _loadCurrentUserAccountsInsideMapping_callback(contractObj);
//     }



  //}
//REVIEW: following is attempt to refactor fetchjson
// async function fetchJSON(JSONid, callback){
//   try {
//           let httpStr = 'https://api.jsonbin.io/b/' + JSONid + '/latest';
//           await fetch(httpStr)
//            .then((response) => response.json())
//            .then((responseJson) => {
//              if(responseJson.length !== 0){
//                 responseJson = checkUndefined(responseJson);
//                 callback(responseJson);
//               // _loadsetRankingListJSONData_callback(responseJson);
//               }
//            })
//         }catch (err) {
//              return console.error(err);
//           }
// }
          //private function not for export 
          function checkUndefined(responseJson){
              let responseDataAsArray = [];
            if(responseJson.length === undefined){
              //on creation of a new user the [] isn't recognized
              //turn the object into an array for use by BSTable
              //responseJson = "[" + responseJson + "]";
              responseDataAsArray[0] = responseJson;
              responseJson = responseDataAsArray;
              return responseJson;
            } else {
              return responseJson;
            }
          }
//}
