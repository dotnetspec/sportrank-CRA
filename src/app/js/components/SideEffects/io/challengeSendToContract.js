import DSportRank from '../../../../../ABIaddress';
//import { getWeb3defaultAccount } from './web3defaultAccount';
import {getWeb3Accounts} from './web3Accounts';
import JSONops from '../../Logic/JSONops'
//this may appear unnecessary but it's done to
//enable mocking:
//This function returns a tx hash:
  export const challengeSendToContract = async (gasEstimate, challengeState, newrankId, user, selectedOpponentName, data) => {
    console.log('gasEstimate', gasEstimate)
    const account = await getWeb3Accounts();
    const challenge = await DSportRank.methods.challenge(challengeState);
    //return await challenge.send({ from: await getWeb3Accounts(), gas: gasEstimate + 100000 });
    try{
          challenge.send({
            from: account[0], gas: gasEstimate + 100000
          })
          .on('transactionHash', function(hash){
              JSONops._updateDoChallengeJSON(newrankId, user, selectedOpponentName, data);
              console.log('hash', hash);
          })
          .on('receipt', function(receipt){
              console.log('receipt', receipt);
          })
          .on('confirmation', function(confirmationNumber, receipt){
            //console.log('confirmationNumber', confirmationNumber);
          })
        }  catch(err){
            console.log('err inside challengeSendToContract')
            console.log(err)
            // remove loading state and show error message
            //setState({ isLoading: false, error: err.message });
            //setIsLoading(false);
            //setError(err.message)
          }
}
