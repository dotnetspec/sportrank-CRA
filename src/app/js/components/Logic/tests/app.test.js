import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import App  from '../App'
import renderer from 'react-test-renderer'
import { shallow, mount } from 'enzyme';
import { stub } from 'sinon';
import { render, fireEvent, cleanup,
   waitForElement, wait} from '@testing-library/react'
import {renderWithRouter} from '../../../utils'
import 'jest-dom/extend-expect'
//since this is a test it's not actually importing the real 'axios' but
//rather the axios in the __mocks__ folder
import axiosMock  from '../../SideEffects/tests/__mocks__/axios'
import web3ioMock  from '../../SideEffects/tests/__mocks__/web3io'
import { _loadsetJSONData, _loadsetRankingListJSONData, getNewRankId, asyncFetch } from '../../SideEffects/io/Jsonio';
import {fetchMock} from 'fetch-mock'
import _loadCurrentUserAccounts from '../../SideEffects/io/web3io';

//NB: There are no 'props' at the <App /> level. Testing using props
//has to take place in the child components

afterEach(cleanup);

//jest.mock('axios');
jest.mock("../../SideEffects/io/Jsonio");
jest.mock('../../SideEffects/io/web3io');

//ensure describe blocks don't overlap
//default approach is RTL unless otherwise specified
describe('<App/> ', () => {

  //default values, override in tests if necessary
  const testAccountPlayer1Rinkeby = '0x847700B781667abdD98E1393420754E503dca5b7';
  const globalRankingData = [{RANKINGNAME: "mplayer1rank", RANKINGDESC: "mp1r", ACTIVE: true, RANKINGID: "5c875c79adeb832d3ec6732d"}]
  const url = '/'
  //Functions:
  const historyMock = { push: jest.fn() };
  const onClick = jest.fn();
  const onAfterUserUpdate = jest.fn();
  const newrankIdCB = jest.fn();
  const viewingOnlyCB = jest.fn();

  const userAccountsArray =
   [
       { address: '0x847700B781667abdD98E1393420754E503dca5b7',
         balance: 2.0,
         user: {
            username: 'player1',
            description: "test2",
            email: "test@test.com",
            owner: "0x847700B781667abdD98E1393420754E503dca5b7",
            picture: "Qmcs96FrhP5N9kJnhNsU87tUsuHpVbaSnGm7nxh13jMLLL",
            rankingDefault: "5c81c1e944e81057efe3e2c8"
         }
       }
     ];

     const userObj = {
               username: 'player1',
               description: "test2",
               email: "test@test.com",
               owner: testAccountPlayer1Rinkeby,
               picture: "Qmcs96FrhP5N9kJnhNsU87tUsuHpVbaSnGm7nxh13jMLLL",
               rankingDefault: "5c81c1e944e81057efe3e2c8"
          };

          const props  = {
            userAccounts: userAccountsArray,
            rankingListJSONdata: globalRankingData,
            account: testAccountPlayer1Rinkeby,
            user: userObj
          }


          xit("App ETH bal basic render w/o data", async () => {
            //override global mockResolvedValue with:
            //axiosMock.get.mockResolvedValueOnce({data: globalRankingData});
            //const mockFn = jest.fn();
            //mockFn.getMockName('_loadCurrentUserAccounts')


            const { getByTestId, getByText, debug, act } = renderWithRouter(<App />);

            //act(() => {
                web3ioMock.get.mockResolvedValue({data: userAccountsArray});
            //});
   /* assert on the output */

            //debug();
              //await wait(() => expect(getByTestId("CurrentETHBal")).toHaveTextContent('SportRank has contributed:'));
            await wait(() => expect('_loadCurrentUserAccounts').toHaveBeenCalled());
          });

      //using the axiosmock file in __mocks__
      xit("App axios mock fetch call", async () => {
        //override global mockResolvedValue with:
        axiosMock.get.mockResolvedValueOnce({data: globalRankingData});
        const { getByTestId, getByText, debug } = renderWithRouter(<App />);
        //debug();
        expect(getByTestId("loading")).toHaveTextContent('Loading ...');
        await wait(() => expect(getByText("mplayer1rank")).toBeInTheDocument());
      });


          xit('Displays De-Activate btn when ranking selected', () => {

            const props  = {
              userAccounts: userAccountsArray,
              isCurrentUserActive: true,
              rankingListJSONdata: globalRankingData,
              account: testAccountPlayer1Rinkeby,
              user: 'player1',
              onChildClick: onClick,
              onAfterUserUpdate:onAfterUserUpdate,
              newrankIdCB:newrankIdCB,
              viewingOnlyCB:viewingOnlyCB,
              history:historyMock,
              data: globalRankingData
            }
                const { getByTestId  } = renderWithRouter(<App url={url}/>);

                const firstRowOfTableViewBtn = getByTestId("0");
                fireEvent.click(firstRowOfTableViewBtn);
                //expect(document.querySelector('[data-testid="activatebtn-input"]')).toBeInTheDocument();
                expect (getByTestId('activatebtn-input')).toBeInTheDocument();
          });

  xit('RTL - check initial display', () => {

        const { getByText, debug  } = renderWithRouter(<App {...props}/>);
        //debug();
        expect(getByText(/List All Rankings/i)).toHaveTextContent('List All Rankings')
        expect (getByText(/Update Profile/i)).toBeInTheDocument();
        expect(document.querySelector('[data-testid="activatebtn-input"]')).not.toBeInTheDocument();
 });

//currently unable to mock the globalRankingData
  xit('loads and displays greeting', async () => {
    //const url = '/'
    const { getByText, getByTestId } = renderWithRouter(<App url={url} />)
    //const globalRankingData = {RANKINGNAME: "mplayer1rank", RANKINGDESC: "mp1r", ACTIVE: true, RANKINGID: "5c875c79adeb832d3ec6732d"}

    axiosMock.get.mockResolvedValueOnce({
      //data: { rankingListData: globalRankingData },
      rankingListData: globalRankingData
    })

    const firstRowOfTableViewBtn = await getByTestId("0");
    fireEvent.click(firstRowOfTableViewBtn);

    //fireEvent.click(getByText('Load Greeting'))

    // const greetingTextNode = await waitForElement(() =>
    //   getByTestId('greeting-text')
    // )

    expect(axiosMock.get).toHaveBeenCalledTimes(1)
    expect(axiosMock.get).toHaveBeenCalledWith(url)
    expect(getByTestId('globalrankingviewbtn')).toHaveTextContent('View')
    //expect(getByTestId('globalrankingviewbtn')).toHaveAttribute('disabled')
  })

//this test causes error: Cannot log after tests are done. Did you forget to wait for something async in your test?
//Attempted to log "Warning: Can't perform a React state update on an unmounted component
// xit('RTL - check elements rendered to DOM when render(<App />) (many e.g. most btns are not!)', () => {
//   const { getByText } = renderWithRouter(<App />);
//   expect (getByText('Sportrank HOME')).toBeInTheDocument();
//   });

//TODO: Need to figure out visibility in child components before coming back to
//how this works at the App level
xit('RTL - check btn visibility', () => {
      // Render new instance in every test to prevent leaking state
      // const historyMock = { push: jest.fn() };
      // const onClick = jest.fn();
      // const onAfterUserUpdate = jest.fn();
      // const newrankIdCB = jest.fn();
      // const viewingOnlyCB = jest.fn();
      //
      // const row = {RANKINGNAME: "mplayer1rank", RANKINGDESC: "mp1r", ACTIVE: true, RANKINGID: "5c875c79adeb832d3ec6732d"}

      //const { getByText, container, queryByTestId } = renderWithRouter(<App />);
      const { queryByTestId, getByTestId, getElementById, rootNode, getByText } = renderWithRouter(<App />);
      //const ancestor = queryByTestId('app')



      //const {queryByTestId} = render(/* Rendered HTML */)
//const ancestor = document.querySelector('[data-testid="app"]')
//const ancestor = queryByTestId(container, 'app')
//const descendant = queryByTestId('header')
//const nonExistantElement = queryByTestId('globalrankingviewbtn')

//expect(document.querySelector('[data-testid="app"]')).toBeInTheDocument()
//expect(document).toContainQuerySelector('[data-testid="app"]')

//expect(getElementById('app')).toBeInTheDocument()
//const root = rootNode.querySelector('.app')
//REVIEW: can be improved but ensures exists ...
expect (getByText(/Update Profile/i)).toHaveTextContent('Update Profile');

//expect(getByTestId('app')).toBeInTheDocument()

// expect(ancestor).toContainElement(descendant)
// expect(descendant).not.toContainElement(ancestor)
// expect(ancestor).not.toContainElement(nonExistantElement)

      //looking for text that hasn't been rendered just causes error
        //expect(getByText("De-Activate")).not.toBeVisible();
        //expect(getByText("De-Activate")).not.toBeVisible();
        //expect(getByText(container, 'De-Activate')).not.toBeVisible()
        //expect(document.querySelector('[data-testid="activatebtn-input"]')).not.toBeVisible()

      // fireEvent.click(getByText(/View/i));
      // expect(onClick).toHaveBeenCalled();
      // expect(onAfterUserUpdate).toHaveBeenCalled();
      // expect(newrankIdCB).toHaveBeenCalled();
      // expect(viewingOnlyCB).toHaveBeenCalled();
    });



    xit("RTL - List All Ranking btn should be visisble", () => {
      // jest.mock('react-router', () => ({
      //   withRouter: Comp => props => <Comp {...props} />,
      // }))
        const { getByTestId, container, getByValue } = renderWithRouter(<App />);
        //expect(container.innerHTML).toMatch('List All Rankings')
          expect(getByValue(container, 'List All Rankings').toHaveTextContent('List All Rankings'))
          //console.log(container.classList, container.id, container.innerHTML);
          //expect(getByTestId('ListAllRankings')).toHaveAttribute('enabled')
        //const container = render(<BrowserRouter><App /></BrowserRouter>)
        //looking for text that hasn't been rendered just causes error
          //expect(getByText("De-Activate")).not.toBeVisible();
          //expect(getByText("List All Rankings")).toBeVisible();
          //expect(document.querySelector('.ListAllRankings').innerHTML).toBe("List All Rankings")
          //expect(getByTestId('ListAllRankings')).toHaveTextContent('List All Rankings')
          //expect(getByText('List All Rankings')).toBeInTheDocument()
      });
});
