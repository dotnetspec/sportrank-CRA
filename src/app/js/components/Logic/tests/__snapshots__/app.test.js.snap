// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App mount <BrowserRouter><App /></BrowserRouter> correctly 1`] = `ReactWrapper {}`;

exports[`App renders whole App.js correctly 1`] = `ShallowWrapper {}`;

exports[`App shallow <BrowserRouter><App /></BrowserRouter> correctly 1`] = `ShallowWrapper {}`;
